src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b);                   // Swap
+print('$a $b');                    // Prints "right left".
+```
+
+### Destrcuturing multiple returns
+
+[Records][] allow aggregating and returning multiple values from a single function
+call. Patterns add the ability to destructure a record's fields
+directly into local variables, inline with the function call.
+
+Instead of individually declaring new local variables for each record field,
+like this:
+
+```dart
+main() {
+  var info = userInfo(json);
+  var name = info.$1;
+  var age = info.$2;
+}
+```
+You can call the function directly into a [variable declaration](#variable-declaration)
+or [assigment pattern](#variable-assignment),
+with a [record subpattern][record] and fields for each value in the return:
+
+```dart
+main() {
+  var (name, age) = userInfo(json);
+}
+```
+
+### Switch cases
+
+The `case` keyword is always followed by a pattern. _Case patterns_ are [refutable][];
+they allow control flow to either match and destructure the object being switched
+on, or continue execution if the object doesn't match. Case patterns can be any
+kind of pattern:
+
+- [**Constant patterns**][constant]: the object's value has to be equivalent to match.
+  ```dart
+  switch (number) {
+    case 1: ...          // Matches if number == 1  
+  }
+  ```
+- [**List**][list], [**map**][map], or [**record**][record] **patterns**: the elements are destructured if matched.
+  ```dart
+  switch (obj) {
+    case [a, b]: ...     // Matches if obj is a list with two elements  
+  }
+  ```  
+- [**Variable patterns**][variable]: the matched value binds to a new variable.
+  ```dart
+  switch ((1, 2)) {
+    case (var a, var b): ...    // Matches if 1 and 2 can be assigned to a and b
+  }
+  ```  
+- [**Logical**][logical] or [**relational**][relational] **patterns**: the object matches if the comparison to
+a constant returns `true`.
+  ```dart
+  switch (code) {
+    case >= first && <= last: ...    
+    // Matches if the value of code is between the constant values of first and last
+  }
+  ```  
+- [**Null-check**][check] or [**null-assert**][assert] **patterns**: the match is first on whether the object is not null, then on the value.
+  ```dart
+  switch (maybe) {
+    case var s?: ... 
+    // Matches if maybe is not null, and then against the inner pattern "var s"
+  }
+  ```  
+- [**Object patterns**][object]: the match exposes getters on the object type.
+  ```dart
+  switch (Shape shape) {
+    case Rect(width: var w, height: var h): ...
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a pattern in a case become local variables,
```

Comment:
I'm not sure if it makes a lot of sense to list out all of the kinds of patterns here when they are also in the other reference page. Also, most of these patterns aren't specific to switch cases, so it arguably makes as much sense to have them here as it would to have them under any of the preceding sections: you can have object, list, and map patterns in pattern variable declarations and pattern assignments too.

Reply:
Ok, good point. I placed this section here (originally from the switch case content) before writing the patterns reference page, but now that that exists this can definitely go